orc8r/cloud/go/services/certifier/servicers/certifier.go

```diff
@@ -712,60 +718,166 @@ func (srv *CertifierServer) getPolicyDecisionFromTokenMany(tokens *certprotos.To
+	case *certprotos.Policy_Network:
+		policyType = constants.NetworkID
+	case *certprotos.Policy_Tenant:
+		policyType = constants.TenantID
+	}
+	return policyType
+}
+
+// doResourceTypesMatch checks if the policy type applies to the requested type
+// for network and tenant resource types.
+// Every request has a path but not necessarily network/tenant ids, so path policy
+// types are always applied to every request, and they are exempted from being matched by type.
+// Returns true if the resource type is either network or tenant and the resource types match
+// Returns false otherwise
+func doResourceTypesMatch(req *certprotos.Request, policy *certprotos.Policy) bool {
+	reqType := getRequestType(req)
+	policyType := getPolicyType(policy)
+	return req.GetResourceId() == nil || policyType == constants.Path || reqType == policyType
+}
+
+// isActionAuthorized checks if the requested action to read/write is authorized by the policy
+// Returns the effect if the policy allows/denies write access regardless of requested action,
+// or if policy allows/denies read access and the requested action is read
+// Otherwise, returns an unknown effect, which implies that the policy does not apply to this requested resource
+func isActionAuthorized(req *certprotos.Request, policy *certprotos.Policy) certprotos.Effect {
 	// The policy does not handle this case, so return UNKNOWN.
-	if reqResource.ResourceType != policyResource.ResourceType {
+	if !doResourceTypesMatch(req, policy) {
 		return certprotos.Effect_UNKNOWN
 	}
-	reqPath := reqResource.Resource
-	policyPath := policyResource.Resource
-	if ok, _ := doublestar.Match(policyPath, reqPath); ok {
-		return policyResource.Effect
+	if policy.Action == certprotos.Action_WRITE {
+		return policy.Effect
+	}
+	if policy.Action == certprotos.Action_READ && req.Action == certprotos.Action_READ {
+		return policy.Effect
 	}
 	return certprotos.Effect_UNKNOWN
 }
 
-// checkAction checks if the requested action is authorized by the policy
-func checkAction(reqResource *certprotos.Resource, policyResource *certprotos.Resource) certprotos.Effect {
+// isResourceAuthorized checks if the requested resource (either a path, network, or tenant) is authorized by the policy
+// Returns the effect if the policy's path matches the requested path and if the policy's tenant/network ID matches that
+// of the requested when applicable
+// Otherwise, returns an unknown effect, which implies that the policy does not apply to this requested resource
+func isResourceAuthorized(req *certprotos.Request, policy *certprotos.Policy) certprotos.Effect {
 	// The policy does not handle this case, so return UNKNOWN.
-	if reqResource.ResourceType != policyResource.ResourceType {
+	if !doResourceTypesMatch(req, policy) {
 		return certprotos.Effect_UNKNOWN
 	}
-	if policyResource.Action == certprotos.Action_WRITE {
-		return policyResource.Effect
+
+	finalEffect := certprotos.Effect_UNKNOWN
+	if path := policy.GetPath(); path != nil {
```

Comment:
"GetPath" will return an empty string if not found, I believe. We probably just want `policy.Path`

Reply:
Was looking at the `GetPath` function in `certifier.pb.go`, I think it does return nil. Also because we were using `oneof` operator in the proto, I believe it is the best way to get the path. The only other way might be `policy.Resource.Path` but because `policy.Resource` resolves to the strange interface function, not sure how to get the path from that.

Reply:
You're totally right. Thanks for looking into this and getting to a thorough understanding for both of us 🙂